brave/brave-core
# New Tab Page WebUI: Use Profile Preferences to store page options instead of localStorage

Fix https://github.com/brave/brave-browser/issues/5064
Fix https://github.com/brave/brave-browser/issues/2848

This is created as a PR on top of master and does not include the extra preferences from https://github.com/brave/brave-core/pull/2762 since it has not been merged yet.

This allows:
- Surfacing these preferences outside of the NTP
- Syncing / migrating the preferences
- Making the change immediately for all open NTP tabs

Use FireWebUIListener to push data to javascript event subscribers.
Still uses render_view_host->SetWebUIProperty (like the existing NTP data), which we may port in the future to a DataSource Dictionary like the other WebUI in chromium.

Provides any data retrieved or updated to redux store via redux action, but does not use redux actions to ask for a data change.

### Also includes

#### Refactor NTP WebUI backend
Focus on the MessageHandler in order to use FireWebUIListener and follow chromium best practice for setting data and events

#### TS types for NTP actions
Small change for intellisense™ and action parameter type checking

## Submitter Checklist:

- Verified that these changes build without errors on
- Verified that these changes pass automated tests (unit, browser, security tests) on

## Test Plan:
1. Open multiple windows with NTP visible in each.
2. Toggle each of the settings
3. Expected: settings change and widgets show/hide in current window and other windows
4. (optional) Go to profile directory, open up the `Default/Preferences` file and look for `brave.new_tab_page.show_background_image` and either true/false (depending on what you set in the UI)

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.
